crypto/cmp/cmp_hdr.c

```diff
@@ -156,6 +156,33 @@ int ossl_cmp_hdr_set1_senderKID(OSSL_CMP_PKIHEADER *hdr,
 {
     if (!ossl_assert(hdr != NULL))
         return 0;
+     if (senderKID == NULL) {
+        /*-
+         * Unprotected, password-based, or protection cert has no subjectKeyIdentifier.
+         * Take senderKID from the commonName in the sender field if available
+         */
+        const GENERAL_NAME *gn = hdr->sender;
+        const X509_NAME *sender;
+        int res;
+
+        if (gn->type == GEN_DIRNAME && (sender = gn->d.directoryName) != NULL
+            && (res = X509_NAME_get_index_by_NID(sender, NID_commonName, -1)) >= 0) {
+            ASN1_STRING *astr = X509_NAME_ENTRY_get_data(X509_NAME_get_entry(sender, res));
+            ASN1_OCTET_STRING *ostr = NULL;
+
+            if (!ossl_cmp_asn1_octet_string_set1_bytes(&ostr, ASN1_STRING_get0_data(astr),
+                                                       ASN1_STRING_length(astr)))
+                return 0;
+            res = ossl_cmp_asn1_octet_string_set1(&hdr->senderKID, ostr);
+            ASN1_OCTET_STRING_free(ostr);
+            return res;
+        }
+    }
+    if (senderKID == NULL) {
+        /* According to RFC 9483 section 3.1, senderKID MUST be used */
```

Comment:
As per RFC, It is only mandatory for password-based protection. In case of signature based protection, it is mandatory  only if corresponding certificate has SubjectKeyIdentifier.